src/minimization.cpp

```diff
@@ -0,0 +1,236 @@
+    if (forwardLayerSweep) {
+        return nodeIndex + portOrderValueDefinition(node, portIndex, forwardLayerSweep) / node.succ.size() + 1;
+    } else {
+        return nodeIndex + portOrderValueDefinition(node, portIndex, forwardLayerSweep) / node.pred.size() + 1;
+    }
+}
+
+void sortNodes(Net & net, std::vector < TreeNode::nodeId > & layer) {
+    for (int i = 0; i < layer.size() - 1; ++i) {
+        for (int j = 0; j < layer.size() - i - 1; ++j) {
+            if (net.getNode(layer[j]) -> barycentricValue > net.getNode(layer[j + 1]) -> barycentricValue) {
+                std::swap(net.getNode(layer[j]) -> number, net.getNode(layer[j + 1]) -> number);
+                std::swap(layer[j], layer[j + 1]);
+            }
+        }
+    }
+}
+
+int getAmountOfLayers(std::vector < TreeNode > & nodes) {
+    int amountOfLayers = 0;
+    for (int i = 0; i < nodes.size(); ++i) {
+        if (nodes[i].layer > amountOfLayers)
+            amountOfLayers = nodes[i].layer;
+    }
+    return amountOfLayers;
+}
+
+std::vector < std::vector < TreeNode::nodeId >> Net::getNodesByLayer() {
+    int amountOfLayers = getAmountOfLayers(this -> nodes);
+    std::vector < std::vector < TreeNode::nodeId >> nodesByLayer(amountOfLayers + 1);
+    for (int i = 0; i < this -> nodes.size(); ++i) {
+        nodesByLayer.at(this -> nodes[i].layer).push_back(this -> nodes[i].id);
+        this -> nodes[i].number = nodesByLayer.at(this -> nodes[i].layer).size() - 1;
+    }
+    return nodesByLayer;
+}
+
+int getPortIndex(TreeNode::nodeId id, TreeNode node, bool forwardLayerSweep) {
+    if (forwardLayerSweep) {
+        for (int i = 0; i < node.succ.size(); ++i) {
+            if (node.succ[i] == id)
+                return i;
+        }
+    } else {
+        for (int i = 0; i < node.pred.size(); ++i) {
+            if (node.pred[i] == id)
+                return i;
+        }
+    }
+}
+
+int crossCounting(Net & net, std::vector < std::vector < TreeNode::nodeId >> & tempNodesByLayer) {
+    int crossNum = 0;
+    for (int i = 1; i < tempNodesByLayer.size(); ++i) {
+        std::vector < std::vector < int >> edges;
+        for (int j = 0; j < tempNodesByLayer[i].size(); ++j) {
+            TreeNode node = * net.getNode(tempNodesByLayer[i][j]);
+            for (int k = 0; k < node.pred.size(); ++k) {
+                if (net.getNode(node.pred[k]) -> layer > node.layer)
+                    continue;
+                std::vector < int > temp(3);
+                temp[0] = net.getNode(node.pred[k]) -> number;
+                temp[1] = j;
+                edges.push_back(temp);
+            }
+            for (int k = 0; k < node.succ.size(); ++k) {
+                if (net.getNode(node.succ[k]) -> layer > node.layer)
+                    continue;
+                std::vector < int > temp(3);
+                temp[0] = net.getNode(node.succ[k]) -> number;
+                temp[1] = j;
+                edges.push_back(temp);
+            }
+        }
```

Comment:
Considering nodes are bound to layers, is it really necessary to recalculate edges every time?